src/dmd/expressionsem.d

```diff
@@ -6774,6 +6774,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                          * Rewrite as:
                          *  e1 = init, e1.ctor(e2)
                          */
+
+                        // Fix Issue 5153 : https://issues.dlang.org/show_bug.cgi?id=5153
+                        // Using `new` to initialize a struct object is a common mistake, but
+                        // the error message from the compiler is not very helpful in that
+                        // case. If exp.e2 is a NewExp and the type of new is the same as
+                        // the type as exp.e1 (struct in this case), then we know for sure
+                        // that the user wants to instantiate a struct. This is done to avoid
+                        // issuing an error when the user actually wants to call a constructor
+                        // which receives a class object.
+                        //
+                        // Foo f = new Foo2(0); is a valid expression if Foo has a constructor
+                        // which receives an instance of a Foo2 class
+                        if (exp.e2.op == TOK.new_)
+                        {
+                            auto newExp = cast(NewExp)(exp.e2);
+                            if (newExp.newtype && newExp.newtype == t1)
+                            {
+                                error(exp.loc, "cannot initialize a `struct` object using `new`. Remove the `new` keyword.");
+                                return setError();
+                            }
+                        }
+
```

Comment:
The error message should also print the offending expression, which will make it easier to find in complex code. Something along the lines of:
```
cannot implicitly convert expression `new Foo(0)` of type `Foo*` to `Foo`, perhaps remove the `new` keyword?
```